cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -445,133 +447,182 @@ func (m *McmManager) SetMachineDeploymentSize(machinedeployment *MachineDeployme
 }
 
 // DeleteMachines deletes the Machines and also reduces the desired replicas of the Machinedeplyoment in parallel.
-func (m *McmManager) DeleteMachines(machines []*Ref) error {
+func (m *McmManager) DeleteMachines(machinesAskedForRemoval []*Ref) error {
```

Comment:
When CA restarts then it calls `cleanUpIfRequired` once and will release all `ToBeDeleted` taints. So now consider the following situation:
All Nodes in a nodegroup: [N1, N2, N3, N4]
1. At T1 CA marked [N2, N4] with `ToBeDeleted` taint. MCM provider added priority=1 and reduced the replicas by 2
2. CA restarts and that calls `cleanUpIfRequired` that removes all the taints. It does not call `DeleteNodes` and therefore `DeleteMachines` is also never called (which is where you are resetting the priorities). In the mean time since both noschedule and tobedeleted taints are removed, workload is scheduled and now MCM goes ahead and deletes the node because we do not check if the `ToBeDeleted` taint is present on a node before deleting the corresponding machine in MCM.

Reply:
As per our offline discussion, I will add it to the PR description